File: app/src/main/java/com/jeeps/gamecollector/model/Game.java
package com.jeeps.gamecollector.model;

import com.google.gson.annotations.SerializedName;

import java.io.Serializable;

/**
 * Created by jeeps on 12/23/2017.
 */
public class Game implements Serializable {
    @SerializedName("gameId")
    private String id;
    //Game data
    private String user;
    private String dateAdded;
    private String imageUri;
    private boolean isPhysical;
    private String name;
    private String shortName;
    private String platformId;
    private String platform;
    private String publisherId;
    private String publisher;
    private int timesCompleted;

    public Game(String imageUri, boolean isPhysical, String name, String shortName,
                String platformId, String platform, String publisherId, String publisher) {
        this.imageUri = imageUri;
        this.isPhysical = isPhysical;
        this.name = name;
        this.shortName = shortName;
        this.platformId = platformId;
        this.platform = platform;
        this.publisherId = publisherId;
        this.publisher = publisher;
        timesCompleted = 0;
    }

    public Game() {}

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getUser() {
        return user;
    }

    public void setUser(String user) {
        this.user = user;
    }

    public String getDateAdded() {
        return dateAdded;
    }

    public void setDateAdded(String dateAdded) {
        this.dateAdded = dateAdded;
    }

    public String getImageUri() {
        return imageUri;
    }

    public void setImageUri(String imageUri) {
        this.imageUri = imageUri;
    }

    public boolean isPhysical() {
        return isPhysical;
    }

    public void setPhysical(boolean physical) {
        isPhysical = physical;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getShortName() {
        return shortName;
    }

    public void setShortName(String shortName) {
        this.shortName = shortName;
    }

    public String getPlatformId() {
        return platformId;
    }

    public void setPlatformId(String platformId) {
        this.platformId = platformId;
    }

    public String getPlatform() {
        return platform;
    }

    public void setPlatform(String platform) {
        this.platform = platform;
    }

    public String getPublisherId() {
        return publisherId;
    }

    public void setPublisherId(String publisherId) {
        this.publisherId = publisherId;
    }

    public String getPublisher() {
        return publisher;
    }

    public void setPublisher(String publisher) {
        this.publisher = publisher;
    }

    public int getTimesCompleted() {
        return timesCompleted;
    }

    public void setTimesCompleted(int timesCompleted) {
        this.timesCompleted = timesCompleted;
    }
}

File: app/src/main/java/com/jeeps/gamecollector/remaster/data/model/data/games/Game.kt
package com.jeeps.gamecollector.remaster.data.model.data.games

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import com.jeeps.gamecollector.remaster.data.model.data.igdb.GameIG
import java.io.Serializable
import java.time.DateTimeException
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/**
 * Created by jeeps on 12/23/2017.
 */

private const val RELEASE_DATE_FORMAT = "MMM dd, yyyy"

@kotlinx.serialization.Serializable
data class Game(
    @SerializedName("gameId")
    val id: String = "",

    //Game data
    val user: String = "",
    val dateAdded: String = "",
    val imageUri: String = "",
    @field:JvmField
    val isPhysical: Boolean = true,
    val name: String = "",
    val shortName: String = "",
    val platformId: String = "",
    val platform: String = "",
    val publisherId: String = "",
    val publisher: String = "",
    val timesCompleted: Int = 0,
    val completionDate: String = "",
    val gameHoursStats: GameHoursStats = GameHoursStats(),

    // AdditionalDetails
    val firstReleaseDate: Long = 0,
    val ageRatings: List<Int> = emptyList(),
    val criticsRating: Double = 0.0,
    val criticsRatingCount: Int = 0,
    val userRating: Double = 0.0,
    val userRatingCount: Int = 0,
    val totalRating: Double = 0.0,
    val totalRatingCount: Int = 0,
    val genres: List<Int> = emptyList(),
    val genresNames: List<String> = emptyList(),
    val storyline: String = "",
    val summary: String = "",
    val url: String = "",

    @Expose(serialize = false, deserialize = false)
    val currentSortStat: String = ""
) : Serializable {

    val completionDateParsed: LocalDate?
        get() = completionDate.ifEmpty { null }?.let {
            Instant.parse(it)
                .atZone(ZoneId.systemDefault())
                .toLocalDate()
        }

    val completionDateFormatted: String
        get() {
            val formatter: DateTimeFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy")
            return completionDateParsed?.format(formatter).orEmpty()
        }
}

fun Game.addAdditionalGameDetails(gameIG: GameIG, genresIg: List<String> = emptyList()): Game {
    return copy(
        firstReleaseDate = gameIG.firstReleaseDate ?: 0,
        ageRatings = gameIG.ageRatings ?: emptyList(),
        criticsRating = gameIG.criticsRating ?: 0.0,
        criticsRatingCount = gameIG.criticsRatingCount ?: 0,
        userRating = gameIG.userRating ?: 0.0,
        userRatingCount = gameIG.userRatingCount ?: 0,
        totalRating = gameIG.totalRating ?: 0.0,
        totalRatingCount = gameIG.totalRatingCount ?: 0,
        genres = gameIG.genres ?: emptyList(),
        genresNames = genresIg,
        storyline = gameIG.storyline ?: "",
        summary = gameIG.summary ?: "",
        url = gameIG.url ?: ""
    )
}

fun Game?.releaseDateFormatted(): String {
    if (this == null || firstReleaseDate == 0L) return ""
    return try {
        val date = Instant.ofEpochSecond(firstReleaseDate)
        val formatter = DateTimeFormatter
            .ofPattern(RELEASE_DATE_FORMAT)
            .withZone(ZoneId.from(ZoneOffset.UTC))
        formatter.format(date)
    } catch (e: DateTimeException) {
        e.printStackTrace()
        ""
    }
}

// Jetpack navigation decodes this slash that breaks the firestore link during serialization
fun Game?.encodeImageUriPath() : Game? {
    if (this == null) return null
    return if (imageUri.contains("/o/gameCovers/")) {
        copy(
            imageUri = imageUri.replace("/o/gameCovers/", "/o/gameCovers%2F")
        )
    } else this
}